src/content/en/fundamentals/media/manipulating/live-effects.md

```diff
@@ -0,0 +1,374 @@
+
+If even if you parallelize your code with `Worker`s, your GPU may have 100s of times as many
+specialized cores as your CPU has general cores.
+
+Examples in this article are mostly for 2D canvas, but the principles should translate pretty easily
+to fragment shader code.
+
+## Effect Types
+
+### Pixel Effects
+This is the simplest category to both understand and implement. All of these transformations take
+the color value of a single pixel and pass it into a function that will give you back another
+color value to use as the output.
+
+There are many variations on these operations that are more or less complicated. Some will take into
+account how the human brain processes visual information based on decades of research, and some will
+be dead simple ideas that give an effect that's mostly reasonable. Either way, the idea is the same.
+
+For example, a brightness control can be implemented by simply taking the red, green and blue values
+of the pixel and multiplying them by a brightness value. A brightness of 0 will make the image
+entirely black. A value of 1 will leave the image unchanged. A value greater than 1 will make it
+brighter.
+
+For 2D canvas:
+
+    const brightness = 1.1; // Make the image 10% brighter
+    for (let i = 0; i < imageData.data.length; i += 4) {
+      imageData.data[i] = imageData.data[i] * brightness;
+      imageData.data[i + 1] = imageData.data[i + 1] * brightness;
+      imageData.data[i + 2] = imageData.data[i + 2] * brightness;
+    }
+
+Note that the loop moves 4 bytes along at a time, but only changes three values - this is because it
+skips changing the alpha value. Also remember that a Uint8ClampedArray will round all values to
+integers, and clamp any value greater than 255.
+
+Some of these filters take extra information, such as the average luminance of the whole
+image, but these are things that can be calculated once for the whole image.
+
+One way of changing contrast, for example, can be to move each pixel towards  or away from some
+'gray' value, for lower or higher contrast respectively. The gray value is commonly chosen to be a
+gray color whose luminance is the median luminance of all pixels in the image.
+
+You can calculate this value once when the image is loaded and then use it each time you need to
+adjust the image effect.
+
+### Multi-pixel
+Some effects use the color of neighboring pixels when deciding the color of the current pixel.
+
+This slightly changes how you do things in the 2D canvas case because you want to be able to read
+the *original* colors of the image, and the previous example was updating the pixels in place.
+
+This is easy enough, though. When you initially create your image data object you can make a copy
+of the data.
+
+    const originalPixels = new Uint8Array(imageData.data);
+
+Note that in the WebGL case it isn't actually possible to query the **new** color of another pixel
+from inside the shader, so you always get the original.
```

Comment:
I stumbled over this. Maybe phrase it from another POV: “Note that you can only access the input image in WebGL, as the output value of other pixels might not have been calculated yet.” or something